cmd/mock-dnf-json/dnf-json.go

```diff
@@ -0,0 +1,120 @@
+// Mock dnf-json
+//
+// The purpose of this program is to return fake but expected responses to
+// dnf-json depsolve and dump queries.  Tests should initialise a
+// dnfjson.Solver and configure it to run this program via the SetDNFJSONPath()
+// method.  This utility accepts queries and returns responses with the same
+// structure as the dnf-json Python script.
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+
+	"github.com/osbuild/osbuild-composer/internal/dnfjson"
+)
+
+func maybeFail(err error) {
+	if err != nil {
+		fail(err)
+	}
+}
+
+func fail(err error) {
+	fmt.Fprintln(os.Stderr, err.Error())
+	os.Exit(1)
+}
+
+func readRequest(data []byte) dnfjson.Request {
+	var req dnfjson.Request
+	maybeFail(json.Unmarshal(data, &req))
+	return req
+}
+
+func respond(result []map[string]interface{}) {
+	resp, err := json.Marshal(result)
+	maybeFail(err)
+	fmt.Printf(string(resp))
+}
+
+func createBaseDepsolveFixture() []dnfjson.PackageSpec {
+	return []dnfjson.PackageSpec{
+		{
+			Name:    "dep-package3",
+			Epoch:   7,
+			Version: "3.0.3",
+			Release: "1.fc30",
+			Arch:    "x86_64",
+			RepoID:  "0",
+		},
+		{
+			Name:    "dep-package1",
+			Epoch:   0,
+			Version: "1.33",
+			Release: "2.fc30",
+			Arch:    "x86_64",
+			RepoID:  "0",
+		},
+		{
+			Name:    "dep-package2",
+			Epoch:   0,
+			Version: "2.9",
+			Release: "1.fc30",
+			Arch:    "x86_64",
+			RepoID:  "0",
+		},
+	}
+}
+
+func generatePackageList() []dnfjson.PackageSpec {
+	var packageList []dnfjson.PackageSpec
+
+	for i := 0; i < 22; i++ {
+		basePackage := dnfjson.PackageSpec{
+			Name:           fmt.Sprintf("package%d", i),
+			Version:        fmt.Sprintf("%d.0", i),
+			Release:        fmt.Sprintf("%d.fc30", i),
+			Arch:           "x86_64",
+			RepoID:         "0",
+			RemoteLocation: fmt.Sprintf("https://pkg%d.example.com", i),
+			Checksum:       fmt.Sprintf("notachecksum-%d", i),
+		}
+
+		secondBuild := basePackage
+
+		secondBuild.Version = fmt.Sprintf("%d.1", i)
+
+		packageList = append(packageList, basePackage, secondBuild)
+	}
+
+	sort.Slice(packageList, func(i, j int) bool {
+		return packageList[i].Name < packageList[j].Name
+	})
+
+	return packageList
+}
+
+func main() {
+	input, err := ioutil.ReadAll(os.Stdin)
+	maybeFail(err)
+
+	req := readRequest(input)
+
+	args := req.Arguments
+	if len(args) == 0 {
+		fail(errors.New("error: empty arguments"))
+	}
```

Comment:
I don't want this test command to do too many things.  It's a replacement for the existing `rpmmd_mock` fixtures.  I actually rewrote it already to be even simpler.  The current tests return fixtures without taking into consideration any of the input and I think we can keep doing the same, otherwise the test becomes a bit too complicated.

One thing to keep in mind, these tests aren't used for testing the new `dnfjson` package itself.  They're for the API tests (mostly used here https://github.com/osbuild/osbuild-composer/blob/main/internal/weldr/api_test.go).  The difference now is that when testing the API, we're still using the real `dnfjson` package but we're mocking the `dnf-json` script, whereas previously we were mocking the entire `rpmmd` package.

Reply:
Thanks for the explanation, makes sense 👍